### ISSUE TITLE
Fix has_package_files when package not installed

### ISSUE THREAD

**Issue by glenjarvis** [glenjarvis/ansible-simple-apt#8]
The reason that python_apt was originally used  was to find if a package has files (there are some that do not). In working through these cases, I seem to have tripped on this original bug.

Stefan Taranu originally gave a suggestion that may work in this case (although it was a dpkg command and the package is not installed). Glen can help find that information.

In short, we are going to bypass this functionality for now as we may just remove that functionality for simplicity. We are recording this ticket to track for now and will return when we have a real-live test case (so we can revisit our options/needs).

We are simply stubbing this out for now and assuming all packages have files until this is better addressed (I have not  yet seen a test case where we haven't had files in packages):

```
    def has_package_files(_ansible_module, package_name, cache_info):
        """Given package_name, return True if package has files"""

        # What happens when package isn't installed?
        #return bool(len(dpkg_files(_ansible_module, package_name)))
        Return True
```

**Comment by glenjarvis** [comment]
To successfully solve this issue, you will need to:

* Deeply understand Debian packaging
* Understand how to write an Ansible module
* Understand the core problem (i.e., why this is happening in a Python virtualenv and not a system python)
* Understand that the "doneness" of this problem is not to "Install in System Python" (that misses the problem we are trying to solve)

To help you see the demo of the problem, several demos have been created using Vagrant so the problem is easily reproduced. There is also the output from a typical run from this demo:

https://github.com/glenjarvis/ansible-simple-apt/tree/develop/demos

**Comment by teizz** [comment]
Going through the original `apt.py` code, reading the comments in the code, and having a fair amount of experience with Debian at this point I think these are the cases `package_status` is trying to cover with regards to `has_files`:

TL;DR: returning the exit-code of `dpkg -L {pkgname} 2>&1 | grep -qE '^/'` will result in the proper legacy behaviour.

Explanation:
* Not-installed/purged packages:
  * Resolve virtual packages if `state == 'install'` and return the correct `is_installed`, `version_ok`, `upgradable` but **always** return `False` for `has_files`.
  * Return `False` for all `installed`, `version_ok`, `upgradable`, and `has_files` otherwise
 * **Removed** packages:
   * Check for non-purged files and return `False` for `is_installed` but `True` for `has_files` if there are files left on the system.
   * The comment says "older python-apt cannot be used to determine non-purged [files]", so always return `False` in that edge case.

Addressing your comment of "I have not yet seen a test case where we haven't had files in packages": I believe the check here is not so much if the original package holds files, rather to check if there are files on the filesystem originally installed by the package being checked.

Having said that, I believe there to be a bug in either python-apt where `apt.Package.installed_files` is a list with a single empty string element (a `[u'']`) if a previously removed package has not been pruned and also does not have any files left behind. At the very least this is a misinterpretation of this by the `apt.py` module as assuming a non-zero length list means `has_files=True`.

Finally, I believe that since `has_files` gets it's information from `installed_files`, and specifically sets it to `False` in case the package is not installed. As well as the fact that listing files from not-installed packages is inherently difficult, that not-installed/purged packages are out of scope with regards to the `has_files` attribute. (The `apt-file` package goes the way of downloading the $repo_Contents-$arch.gz files from the actual deb sources and zgrepping _all_ the files matching with the package being listed, illustrating the length one needs to go through to list all files in a package).

I've included the terminal output of where I:
* start a clean debian:jessie container
* install python + python-apt
* paste the relevant python-apt code to run a test as `python-apt-has-files.py`
* show the `python-apt-has-files.py` snippet to correctly return `False` for packages `python-twisted-core` and `python-twisted`
* show `python-twisted` to be not-installed
* install `python-twisted` which installs `python-twisted-core` amongst others as a dependency
* show `python-twisted-*` packages to be installed
* show `python-apt-has-files.py` correctly returning `True` for `has_files` but already including a `u''` in the list of files
* show the correct number of files listed by `dpkg -L $package`
* autoremove `python-twisted` which removes most dependencies entirely
* show `python-twisted` and `python-twisted-core` left behind with "rc" status
* show `python-apt-has-files.py` incorrectly returning that `python-twisted` has files remaining
* show `dpkg -L` returning the correct number of files as well as a message that `python-twisted` "does not contain any files (!)"

[has-files-tests.txt](https://github.com/glenjarvis/ansible-simple-apt/files/5860301/has-files-tests.txt)

Hope this helps.

**Comment by glenjarvis** [comment]
This was an incredibly good PR. Your testing and information was perfect.

I'm sorry for the delay in merging it. I had meant to spend some time going through a testing suite and reviewing more. However, as I know this has a Rysolv bounty on it, I wanted to get it merged as soon as possible.

I've merged and the issue this fixes is closed.  @teizz  You should receive the Rysolv Bounty for this for such a good fix.

**Comment by teizz** [comment]
Admittedly, I did notice the existence of Rysolv on the frontpage of HN and saw this issue on Rysolv. I figured I could give some feedback, and after doing so figured I might as well drop you a PR for it.

I specifically did not go through Rysolv as I'm not interested in the bounty 😉  If you know who funded it, kindly have them put it towards another cause. If not claiming it defaults it going to Rysolv, I'm also find with that since they did facilitate me finding this.

Either way, I'm happy you're satisfied and are able to close a nearly 5 month old issue 😃.
Cheers!

**Comment by glenjarvis** [comment]
@teizz  I funded the Rysolv myself. So, no worries there I wanted to support them. But, I also really wanted to get in touch with other developers who were technical advanced to tackle a problem that I'm trying to solve.

The core problem is that there several libraries that will not run in a Python virtualenv. And, I'm looking for solutions to fix that -- possibly a forked library and/or simpler replacement library.

Your help was great.
